ENGAGEHF/Dashboard/Vitals/VitalsManager.swift

```diff
@@ -0,0 +1,353 @@
+            quantityType = HKQuantityType(.heartRate)
+        default:
+            throw VitalsError.invalidObservationType
+        }
+        
+        let (startDate, endDate) = observation.getEffectiveDate()
+        
+        guard let startDate, let endDate else {
+            throw VitalsError.invalidConversion
+        }
+        
+        return HKQuantitySample(type: quantityType, quantity: hkQuantity, start: startDate, end: endDate)
+    }
+    
+    private func convertToHKCorrelation(_ observation: R4Observation) throws -> HKCorrelation {
+        // For now, only handle Blood Pressure
+        guard let observationType = observation.code.coding?.first?.code?.value?.string,
+              observationType == "85354-9" else {
+            throw VitalsError.invalidObservationType
+        }
+        
+        let (startDate, endDate) = observation.getEffectiveDate()
+        
+        guard let startDate, let endDate else {
+            throw VitalsError.invalidConversion
+        }
+        
+        // Index into the components of the observation for systolic and diastolic measurements
+        guard let components = observation.component else {
+            throw VitalsError.missingField
+        }
+        
+        let systolicComponent = try self.getComponent(components, code: "8480-6")
+        let diastolicComponent = try self.getComponent(components, code: "8462-4")
+        
+        let systolicQuantity = try self.getQuantity(component: systolicComponent)
+        let diastolicQuantity = try self.getQuantity(component: diastolicComponent)
+        
+        let systolicSample = HKQuantitySample(
+            type: HKQuantityType(.bloodPressureSystolic),
+            quantity: systolicQuantity,
+            start: startDate,
+            end: endDate
+        )
+        let diastolicSample = HKQuantitySample(
+            type: HKQuantityType(.bloodPressureDiastolic),
+            quantity: diastolicQuantity,
+            start: startDate,
+            end: endDate
+        )
+        
+        return HKCorrelation(
+            type: HKCorrelationType(.bloodPressure),
+            start: startDate,
+            end: endDate,
+            objects: [systolicSample, diastolicSample]
+        )
+    }
+    
+    
+    private func getComponent(_ components: [ObservationComponent], code: String) throws -> ObservationComponent {
+        guard let component = components.first(
+            where: {
+                $0.code.coding?.contains(
+                    where: {
+                        $0.code?.value?.string == code
+                    }
+                ) ?? false
+            }
+        ) else {
+            throw VitalsError.missingField
+        }
+        
+        return component
+    }
+    
+    // TODO: Combine the two versions of this into a generic function?
+    private func getQuantity(observation: R4Observation) throws -> HKQuantity {
```

Comment:
We could combine these two functions by introducing a new protocol that `Observation` and `ObservationComponent` both conform to that contains a `value` property returning the value. This way you could pass both into this function that expects an `ObservationComponent` instance.

Even better, we could make this function an extension on the `ObservationComponent` protocol that returns a `quantity` value and this will work for Observations and Observation Components.